A72-open-telemetry-tracing.md

```diff
@@ -0,0 +1,560 @@
+  // base64 encoding/decoding between API boundaries.
+  @Override
+  void set(Metadata header, String key, byte[] value) {
+    assert key.equals("grpc-trace-bin");
+    header.put(Metadata.Key.of(key, BINARY_BYTE_MARSHALLER), value);
+  }
+
+  @Override
+  void set(Metadata header, String key, String value) {
+    if (key.equals("grpc-trace-bin")) {
+      // Slower path. It shows the decoding part of the just encoded
+      // String at GrpcTraceBinPropagator.inject(). This can be used to
+      // propagate any other binary header.
+      header.put(Metadata.Key.of(key, BINARY_BYTE_MARSHALLER), Base64.getDecoder().decode(value));
+    } else if (key.endsWith("-bin")) {
+      logger.log(Level.ERROR, "Binary propagator other than GrpcTraceBinPropagator is not supported.");
+    } else {
+      // Used by other TextMap propagators, e.g. W3C.
+      header.put(Metadata.Key.of(key, ASCII_STRING_MARSHALLER), value);
+    }
+  }
+}
+
+class GrpcCommonGetter implements TextMapGetter<Metadata> {
+  @Override
+  public String get(@Nullable Metadata carrier, String key) {
+    if (key.equals("grpc-trace-bin")) {
+      // Slow path: return string encoded from bytes. Later we decode to
+      // bytes in GrpcTraceBinPropagator.extract().
+      byte[] value = carrier.get(Metadata.Key.of(key, BINARY_BYTE_MARSHALLER));
+      return Base64.getEncoder().encodeToString(value);
+    } else if (key.endsWith("-bin")) {
+      logger.log(Level.ERROR, "Binary propagator other than GrpcTraceBinPropagator is not supported.");
+    } else {
+      // Used by other TextMap propagators, e.g. W3C.
+      return carrier.get(Metadata.Key.of(key, ASCII_STRING_MARSHALLER));
+    }
+  }
+
+  // Add a new method to optimize the TextMap propagator to avoid base64 encoding.
+  @Override
+  public byte[] getBinary(@Nullable Metadata carrier, String key) {
+    assert key.equals("grpc-trace-bin");
+    return carrier.get(Metadata.Key.of(key, BINARY_BYTE_MARSHALLER));
+  }
+}
+
+// This interface will be implemented by gRPCCommonSetter/Getter as an optimization
+// path to avoid base64 encoding between TextMap APIs due to lack of
+// OpenTelemetry binary propagator API. Not for C++.
+private interface GrpcBinarySetter<Metadata> {
+  void set(Metadata header, String key, byte[] value);
+}
+
+```
+
+The `GrpcCommonSetter` adds an overloaded `set()` method to directly take `byte[]` 
+(Java and Go) to avoid extra base64 encoding. For the normal `set()` method it 
+should handle both binary (`-bin`) header and ASCII header from any TextMap 
+propagators that users may configure.
+The `GrpcCommonGetter` in Java and Go adds a new method `getBinary()` for the 
+optimized path for the same reason. Similarly, the normal `get()` method handles
+both binary headers and TextMap propagators.
+
+#### Context Propagation APIs in C++
+C++ will also support propagator APIs to provides API uniformity among
+languages. Since gRPC C++ avoids RTTI, it can not use the same optimization path
```

Comment:
This is the first time we are referring to Run-time type information. Can we add it along with acronym?